src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt

```diff
@@ -0,0 +1,152 @@
+        this{ this.closure() }
+    })
+}
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofSourceSet(sourceSet: String): Collection<GenerateProtoTask> =
+    if (this is ProtobufConfigurator.JavaGenerateProtoTaskCollection)
+        this.ofSourceSet(sourceSet) else emptyList()
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofFlavor(flavor: String): Collection<GenerateProtoTask> =
+    if (this is ProtobufConfigurator.AndroidGenerateProtoTaskCollection)
+        this.ofFlavor(flavor) else emptyList()
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofBuildType(buildType: String): Collection<GenerateProtoTask> =
+    if (this is ProtobufConfigurator.AndroidGenerateProtoTaskCollection)
+        this.ofBuildType(buildType) else emptyList()
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofVariant(variant: String): Collection<GenerateProtoTask> =
+    if (this is ProtobufConfigurator.AndroidGenerateProtoTaskCollection)
+        this.ofVariant(variant) else emptyList()
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofNonTest(): Collection<GenerateProtoTask> =
+    if (this is ProtobufConfigurator.AndroidGenerateProtoTaskCollection)
+        this.ofNonTest() else emptyList()
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofTest(): Collection<GenerateProtoTask> =
+    if (this is ProtobufConfigurator.AndroidGenerateProtoTaskCollection)
+        this.ofTest() else emptyList()
+
+/**
+ * The 'protobuf' configuration.
+ */
+val ConfigurationContainer.protobuf: Configuration
+    get() = getByName("protobuf")
+
+fun DependencyHandler.protobuf(dependencyNotation: Any): Dependency? =
+    add("protobuf", dependencyNotation)
+
+inline fun DependencyHandler.protobuf(
+    dependencyNotation: String,
+    dependencyConfiguration: ExternalModuleDependency.() -> Unit
+): ExternalModuleDependency =
+    add("protobuf", dependencyNotation, dependencyConfiguration)
+
+fun DependencyHandler.protobuf(
+    group: String,
+    name: String,
+    version: String? = null,
+    configuration: String? = null,
+    classifier: String? = null,
+    ext: String? = null
+): ExternalModuleDependency =
+    create(group, name, version, configuration, classifier, ext).apply { add("protobuf", this) }
+
+inline fun DependencyHandler.protobuf(
+    group: String,
+    name: String,
+    version: String? = null,
+    configuration: String? = null,
+    classifier: String? = null,
+    ext: String? = null,
+    dependencyConfiguration: ExternalModuleDependency.() -> Unit
+): ExternalModuleDependency =
+    add("protobuf", create(group, name, version, configuration, classifier, ext), dependencyConfiguration)
+
+inline fun <T : ModuleDependency> DependencyHandler.protobuf(dependency: T, dependencyConfiguration: T.() -> Unit): T =
+    add("protobuf", dependency, dependencyConfiguration)
+
+/**
+ * The 'testProtobuf' configuration.
+ */
+val ConfigurationContainer.testProtobuf: Configuration
```

Comment:
Can this be more general? It looks like this gives special treatment to the main sourceset (`protobuf`) and the test sourceset (`testProtobuf`) but won't handle user created sourcesets like this:


```groovy
sourceSets {
    main {
    }
    foo {
        java {
            srcDirs 'src/foo/'
        }
    }
}


dependencies {
  fooProtobuf files("/tmp/protos-test.tar.gz")
}
```

`createConfiguration` is where the plugin creates the `xxxxProtobuf` configurations in the project: 
https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L154

Maybe there is some way this extension can plug into them in a general way.

Reply:
So I cant dynamically provide the dependency configuration methods, but I can try providing a nicer mechanism. 

You can try it out in the latest commit.
```
dependencies{
    protobuf["sample"]("com.google.protobuf:protobuf-java:3.6.1")
}
```

Reply:
This solution has some pros and cons so I wanted to discuss them with you when possible. Ill add a full explanation in a few.